docs/spec/namedtuples.rst

```diff
@@ -0,0 +1,147 @@
+a named tuple, type checkers should synthesize a ``__new__`` method based on
+the named tuple fields. This mirrors the runtime behavior. In the example
+above, the synthesized ``__new__`` method would look like the following::
+
+    def __new__(cls, x: int, y: int, units: str = "meters") -> None:
+        ...
+
+The runtime implementation of ``NamedTuple`` enforces that fields with default
+values must come after fields without default values. Type checkers should
+likewise enforce this restriction::
+
+    class Location(NamedTuple):
+        altitude: float = 0.0
+        latitude: float  # Type error (previous field has a default value)
+        longitude: float
+
+A named tuple class can be subclassed, but any fields added by the subclass
+are not considered part of the named tuple type. Type checkers should enforce
+that these newly-added fields do not conflict with the named tuple fields
+in the base class::
+
+    class PointWithName(Point):
+        name: str  # OK
+        x: int  # Type error (invalid override of named tuple field)
+
+In Python 3.11 and newer, the class syntax supports generic named tuple classes.
+Type checkers should support this::
+
+    class Property[T](NamedTuple):
+        name: str
+        value: T
+
+    reveal_type(Property("height", 3.4))  # Revealed type is Property[float]
+
+``NamedTuple`` does not support multiple inheritance. Type checkers should
+enforce this limitation::
+
+    class Unit(NamedTuple, object):  # Type error
+        name: str
+
+The function syntax supports two variants: ``collections.namedtuple`` and
+``typing.NamedTuple``. The latter provides a way to specify the types
+of the fields in the tuple whereas the former does not. The ``namedtuple``
+form allows fields to be specified as a tuple or list of strings or a single
+string with fields separated by whitespace or commas. The ``NamedTuple``
+functional form accepts an iterable of ``(name, type)`` pairs.
+For the ``namedtuple`` form, all fields are assumed to be of type ``Any``.
+
+A type checker may support the function syntax in its various forms::
+
+    Point1 = namedtuple('Point1', ['x', 'y'])
+    Point2 = namedtuple('Point2', ('x', 'y'))
+    Point3 = namedtuple('Point3', 'x y')
+    Point4 = namedtuple('Point4', 'x, y')
+
+    Point5 = NamedTuple('Point5', [('x', int), ('y', int)])
+    Point6 = NamedTuple('Point6', (('x', int), ('y', int)))
+
+At runtime, the ``namedtuple`` function disallows field names that are
+illegal Python identifiers and either raises an exception or replaces these
+fields with a parameter name of the form ``_N``. The behavior depends on
+the value of the ``rename`` argument. Type checkers may replicate this
+behavior statically::
+
+    NT1 = namedtuple("NT1", ["a", "b"])  # Type error (duplicate field name)
+    NT2 = namedtuple("NT2", ["abc", "def"], rename=False)  # Type error (illegal field name)
+
+    NT3 = namedtuple("NT3", ["abc", "def"], rename=True)  # OK
+    NT3(abc="", _1="")  # OK
```

Comment:
`def` is not a valid identifier in Python because it's a reserved keyword. If you try running this code, you'll see that a runtime exception is raised.